From: Daniel Golle <daniel@makrotopia.org>
Date: Tue, 16 Sep 2025 22:37:27 +0100
Subject: serial: 8250_mtk: Enable baud clock and manage in runtime PM
Git-commit: d518314a1fa4e980a227d1b2bda1badf433cb932
Patch-mainline: v6.18-rc3
References: git-fixes

Some MediaTek SoCs got a gated UART baud clock, which currently gets
disabled as the clk subsystem believes it would be unused. This results in
the uart freezing right after "clk: Disabling unused clocks" on those
platforms.

Request the baud clock to be prepared and enabled during probe, and to
restore run-time power management capabilities to what it was before commit
e32a83c70cf9 ("serial: 8250-mtk: modify mtk uart power and clock
management") disable and unprepare the baud clock when suspending the UART,
prepare and enable it again when resuming it.

Fixes: e32a83c70cf9 ("serial: 8250-mtk: modify mtk uart power and clock management")
Fixes: b6c7ff2693ddc ("serial: 8250_mtk: Simplify clock sequencing and runtime PM")
Cc: stable <stable@kernel.org>
Reviewed-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Signed-off-by: Daniel Golle <daniel@makrotopia.org>
Link: https://patch.msgid.link/de5197ccc31e1dab0965cabcc11ca92e67246cf6.1758058441.git.daniel@makrotopia.org
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Acked-by: Jiri Slaby <jslaby@suse.cz>
---
 drivers/tty/serial/8250/8250_mtk.c |   11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

--- a/drivers/tty/serial/8250/8250_mtk.c
+++ b/drivers/tty/serial/8250/8250_mtk.c
@@ -429,6 +429,7 @@ static int __maybe_unused mtk8250_runtim
 	if (data->clk_count == 0U) {
 		dev_dbg(dev, "%s clock count is 0\n", __func__);
 	} else {
+		clk_disable_unprepare(data->uart_clk);
 		clk_disable_unprepare(data->bus_clk);
 		data->clk_count--;
 	}
@@ -450,6 +451,12 @@ static int __maybe_unused mtk8250_runtim
 			dev_warn(dev, "Can't enable bus clock\n");
 			return err;
 		}
+		err = clk_prepare_enable(data->uart_clk);
+		if (err) {
+			dev_warn(dev, "Can't enable uart clock\n");
+			clk_disable_unprepare(data->bus_clk);
+			return err;
+		}
 		data->clk_count++;
 	}
 
@@ -484,13 +491,13 @@ static int mtk8250_probe_of(struct platf
 	int dmacnt;
 #endif
 
-	data->uart_clk = devm_clk_get(&pdev->dev, "baud");
+	data->uart_clk = devm_clk_get_enabled(&pdev->dev, "baud");
 	if (IS_ERR(data->uart_clk)) {
 		/*
 		 * For compatibility with older device trees try unnamed
 		 * clk when no baud clk can be found.
 		 */
-		data->uart_clk = devm_clk_get(&pdev->dev, NULL);
+		data->uart_clk = devm_clk_get_enabled(&pdev->dev, NULL);
 		if (IS_ERR(data->uart_clk)) {
 			dev_warn(&pdev->dev, "Can't get uart clock\n");
 			return PTR_ERR(data->uart_clk);
